From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Denis Grigorev <d.grigorev@omprussia.ru>
Date: Wed, 17 Jun 2026 12:55:21 +0200
Subject: Add a video decoder based on gecko-camera. JB#56755

Add the Gecko integration for the package-level gecko-camera decoder and
link its sources into the media platform build.

Initialize and register it in the default, RDD, and content process paths.
Keep remote content decoders ahead of the local hardware fallback.
---
 dom/media/platforms/PDMFactory.cpp       | 33 ++++++++++++++++++++++++
 dom/media/platforms/gecko-camera         |  1 +
 dom/media/platforms/moz.build            | 11 ++++++++
 modules/libpref/init/StaticPrefList.yaml | 13 ++++++++++
 4 files changed, 58 insertions(+)
 create mode 120000 dom/media/platforms/gecko-camera

diff --git a/dom/media/platforms/PDMFactory.cpp b/dom/media/platforms/PDMFactory.cpp
index 6af74f901df0..e44e10c1d59a 100644
--- a/dom/media/platforms/PDMFactory.cpp
+++ b/dom/media/platforms/PDMFactory.cpp
@@ -53,6 +53,9 @@
 #ifdef MOZ_OMX
 #  include "OmxDecoderModule.h"
 #endif
+#ifdef MOZ_ENABLE_WEBRTC_GECKOCAMERA
+#  include "gecko-camera/GeckoCameraDecoderModule.h"
+#endif
 #include <functional>
 
 #include "FFVPXRuntimeLinker.h"
@@ -105,6 +108,9 @@ class PDMInitializer final {
     }
 #endif
     FFVPXRuntimeLinker::Init();
+#ifdef MOZ_ENABLE_WEBRTC_GECKOCAMERA
+    GeckoCameraDecoderModule::Init();
+#endif
   }
 
   static void InitUtilityPDMs() {
@@ -161,6 +167,9 @@ class PDMInitializer final {
     }
 #endif  // !defined(MOZ_WIDGET_ANDROID)
 
+#ifdef MOZ_ENABLE_WEBRTC_GECKOCAMERA
+    GeckoCameraDecoderModule::Init();
+#endif
     RemoteMediaManagerChild::Init();
   }
 
@@ -177,6 +186,9 @@ class PDMInitializer final {
     FFVPXRuntimeLinker::Init();
 #ifdef MOZ_FFMPEG
     FFmpegRuntimeLinker::Init();
+#endif
+#ifdef MOZ_ENABLE_WEBRTC_GECKOCAMERA
+    GeckoCameraDecoderModule::Init();
 #endif
   }
 
@@ -581,6 +593,12 @@ void PDMFactory::CreateRddPDMs() {
     mFailureFlags += GetFailureFlagBasedOnFFmpegStatus(
         FFmpegRuntimeLinker::LinkStatusCode());
   }
+#endif
+#ifdef MOZ_ENABLE_WEBRTC_GECKOCAMERA
+  if (StaticPrefs::media_gecko_camera_codec_enabled()) {
+    StartupPDM(GeckoCameraDecoderModule::Create(),
+               StaticPrefs::media_gecko_camera_codec_preferred());
+  }
 #endif
   StartupPDM(AgnosticDecoderModule::Create(),
              StaticPrefs::media_prefer_non_ffvpx());
@@ -667,6 +685,14 @@ void PDMFactory::CreateContentPDMs() {
   }
 #endif
 
+#ifdef MOZ_ENABLE_WEBRTC_GECKOCAMERA
+  if (StaticPrefs::media_gecko_camera_codec_enabled()) {
+    // Keep remote decoders first, but prefer gecko-camera over local software
+    // decoder modules when content-process fallback is used.
+    StartupPDM(GeckoCameraDecoderModule::Create());
+  }
+#endif
+
 #if !defined(MOZ_WIDGET_ANDROID)  // Still required for video?
   if (StaticPrefs::media_allow_audio_non_utility()) {
 #endif  // !defined(MOZ_WIDGET_ANDROID)
@@ -762,6 +788,13 @@ void PDMFactory::CreateDefaultPDMs() {
   }
 #endif
 
+#ifdef MOZ_ENABLE_WEBRTC_GECKOCAMERA
+  if (StaticPrefs::media_gecko_camera_codec_enabled()) {
+    StartupPDM(GeckoCameraDecoderModule::Create(),
+               StaticPrefs::media_gecko_camera_codec_preferred());
+  }
+#endif
+
   StartupPDM(AgnosticDecoderModule::Create(),
              StaticPrefs::media_prefer_non_ffvpx());
 
diff --git a/dom/media/platforms/gecko-camera b/dom/media/platforms/gecko-camera
new file mode 120000
index 000000000000..215967f59e8c
--- /dev/null
+++ b/dom/media/platforms/gecko-camera
@@ -0,0 +1 @@
+../../../../gecko-camera-decoder
\ No newline at end of file
diff --git a/dom/media/platforms/moz.build b/dom/media/platforms/moz.build
index 1bd4af56f2a2..cc81530596a4 100644
--- a/dom/media/platforms/moz.build
+++ b/dom/media/platforms/moz.build
@@ -130,6 +130,17 @@ if CONFIG["MOZ_WIDGET_TOOLKIT"] == "android":
         "/media/libyuv/libyuv/include",
     ]
 
+if CONFIG["MOZ_ENABLE_WEBRTC_GECKOCAMERA"]:
+    EXPORTS += [
+        "gecko-camera/GeckoCameraDecoderModule.h",
+    ]
+    UNIFIED_SOURCES += [
+        "gecko-camera/GeckoCameraDecoderModule.cpp",
+        "gecko-camera/GeckoCameraVideoDecoder.cpp",
+    ]
+    OS_LIBS += CONFIG["LIBGECKOCAMERA_LIBS"]
+    CXXFLAGS += CONFIG["LIBGECKOCAMERA_CFLAGS"]
+
 FINAL_LIBRARY = "xul"
 
 # Add libFuzzer configuration directives
diff --git a/modules/libpref/init/StaticPrefList.yaml b/modules/libpref/init/StaticPrefList.yaml
index 722d74c6ebef..a74fd0a8b160 100644
--- a/modules/libpref/init/StaticPrefList.yaml
+++ b/modules/libpref/init/StaticPrefList.yaml
@@ -12578,6 +12578,19 @@
     mirror: always
 #endif  # ANDROID
 
+#ifdef MOZ_EMBEDLITE
+  # Enable the gecko-camera based decoder by default.
+-   name: media.gecko-camera-codec.enabled
+    type: RelaxedAtomicBool
+    value: true
+    mirror: always
+
+-   name: media.gecko-camera-codec.preferred
+    type: RelaxedAtomicBool
+    value: true
+    mirror: always
+#endif  # MOZ_EMBEDLITE
+
 # Now we will completely disable the ability to perform audio decoding outside
 # of Utility.
 -   name: media.allow-audio-non-utility
